libcontainer/cgroups/fs/rdma.go

```diff
@@ -0,0 +1,118 @@
+// +build linux
+
+package fs
+
+import (
+	"github.com/opencontainers/runc/libcontainer/cgroups"
+	"github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
+	"github.com/opencontainers/runc/libcontainer/configs"
+	"math"
+	"strconv"
+	"strings"
+)
+
+type RdmaGroup struct {
+}
+
+func (s *RdmaGroup) Name() string {
+	return "rdma"
+}
+
+func (s *RdmaGroup) Apply(path string, d *cgroupData) error {
+	return join(path, d.pid)
+}
+
+func createCmdString(device string, limits configs.LinuxRdma) string {
+	var cmdString string
+	cmdString = device
+	if limits.HcaHandles != nil {
+		cmdString = cmdString + " " + "hca_handle=" + strconv.FormatUint(uint64(*limits.HcaHandles), 10)
+	}
+	if limits.HcaObjects != nil {
+		cmdString = cmdString + " " + "hca_object=" + strconv.FormatUint(uint64(*limits.HcaObjects), 10)
+	}
+	return cmdString
+}
+
+func (s *RdmaGroup) Set(path string, cgroup *configs.Cgroup) error {
+	for device, limits := range cgroup.Resources.Rdma {
+		if err := fscommon.WriteFile(path, "rdma.max", createCmdString(device, limits)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *RdmaGroup) GetStats(path string, stats *cgroups.Stats) error {
+	if !cgroups.PathExists(path) {
+		return nil
+	}
+	currentData, err := fscommon.GetCgroupParamString(path, "rdma.current")
+	if err != nil {
+		return err
+	}
+	currentPerDevices := strings.Split(string(currentData), "\n")
+	maxData, err := fscommon.GetCgroupParamString(path, "rdma.max")
+	if err != nil {
+		return err
+	}
+	maxPerDevices := strings.Split(string(maxData), "\n")
+	// If device got removed between reading two files, ignore returning
+	// stats.
+	if len(currentPerDevices) != len(maxPerDevices) {
+		return nil
+	}
+	currentEntries := ConvertRdmaEntry(currentPerDevices)
+	maxEntries := ConvertRdmaEntry(maxPerDevices)
+
+	stats.RdmaStats = cgroups.RdmaStats{
+		RdmaLimit:   maxEntries,
+		RdmaCurrent: currentEntries,
+	}
+
+	return nil
+}
+
+// Parse raw string to RdmaEntry
+func ParseRdmaKV(raw string, entry *cgroups.RdmaEntry) {
+	var value uint64
+	var err error
+
+	parts := strings.Split(raw, "=")
+	switch len(parts) {
+	case 2:
+		if parts[1] == "max" {
+			value = math.MaxUint32
+		} else {
+			value, err = strconv.ParseUint(parts[1], 10, 32)
+			if err != nil {
+				return
+			}
+		}
+		if parts[0] == "hca_handle" {
+			entry.HcaHandles = uint32(value)
+		} else if parts[0] == "hca_object" {
+			entry.HcaObjects = uint32(value)
+		}
+	}
+}
+
+// Converts array of rawstrings to RdmaEntries
+func ConvertRdmaEntry(strEntries []string) []cgroups.RdmaEntry {
+	var rdmaEntries []cgroups.RdmaEntry
+	for i := range strEntries {
+		parts := strings.Fields(strEntries[i])
```

Comment:
`strings.Fields` is 2x slower than `strings.Split`. Any justification for using it here?

Reply:
@kolyshkin `string.Fields` handles whitespaces very well with appropriate trimming where necessary. With `strings.Split` we would still handle those cases for safety. So i thought since input is a very small constant `Fields` would rather keep code clean without taking any significant hit on performance. But sure if you think `strings.Split` makes more sense i'd rather switch to that. Not a big issue.

Reply:
If the input is coming from the kernel (rather than e.g. a user-written config file), it does not contain any extra spaces (and if it does, something is really wrong). So I'd use strings.Split here.

Reply:
@kolyshkin Sure makes sense. I have upated this and converted to `strings.SplitN`. This is resolved.